src/routes/v3/oauth_clients.rs

```diff
@@ -0,0 +1,387 @@
+use std::{collections::HashSet, fmt::Display};
+
+use actix_web::{
+    delete, get, patch, post,
+    web::{self},
+    HttpRequest, HttpResponse,
+};
+use chrono::Utc;
+use itertools::Itertools;
+use rand::{distributions::Alphanumeric, Rng, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use serde::{Deserialize, Serialize};
+use sqlx::PgPool;
+use validator::Validate;
+
+use super::ApiError;
+use crate::{
+    auth::get_user_from_headers,
+    database::{
+        models::{
+            generate_oauth_client_id, generate_oauth_redirect_id,
+            oauth_client_authorization_item::OAuthClientAuthorization,
+            oauth_client_item::{OAuthClient, OAuthRedirectUri},
+            DatabaseError, OAuthClientId, User,
+        },
+        redis::RedisPool,
+    },
+    models::{self, oauth_clients::OAuthClientCreationResult, pats::Scopes},
+    queue::session::AuthQueue,
+    routes::v2::project_creation::CreateError,
+    util::validate::validation_errors_to_string,
+};
+
+use crate::database::models::oauth_client_item::OAuthClient as DBOAuthClient;
+
+pub fn config(cfg: &mut web::ServiceConfig) {
+    cfg.service(oauth_client_create);
+    cfg.service(oauth_client_edit);
+    cfg.service(oauth_client_delete);
+    cfg.service(get_user_clients);
+    cfg.service(get_user_oauth_authorizations);
+    cfg.service(revoke_oauth_authorization);
+}
+
+#[get("user/{user_id}/oauth_apps")]
+pub async fn get_user_clients(
+    req: HttpRequest,
+    info: web::Path<String>,
+    pool: web::Data<PgPool>,
+    redis: web::Data<RedisPool>,
+    session_queue: web::Data<AuthQueue>,
+) -> Result<HttpResponse, ApiError> {
+    let current_user = get_user_from_headers(
+        &req,
+        &**pool,
+        &redis,
+        &session_queue,
+        Some(&[Scopes::OAUTH_CLIENT_READ]),
+    )
+    .await?
+    .1;
+
+    let target_user = User::get(&info.into_inner(), &**pool, &redis).await?;
+
+    if let Some(target_user) = target_user {
+        let target_user_id: models::ids::UserId = target_user.id.into();
+        current_user.validate_can_interact_with_oauth_client(target_user_id)?;
+
+        let clients = OAuthClient::get_all_user_clients(target_user.id, &**pool).await?;
+
+        let response = clients
+            .into_iter()
+            .map(models::oauth_clients::OAuthClient::from)
+            .collect_vec();
+
+        Ok(HttpResponse::Ok().json(response))
+    } else {
+        Ok(HttpResponse::NotFound().body(""))
+    }
+}
+
+#[derive(Deserialize, Validate)]
+pub struct NewOAuthApp {
+    #[validate(
+        custom(function = "crate::util::validate::validate_name"),
+        length(min = 3, max = 255)
+    )]
+    pub name: String,
+
+    #[validate(
+        custom(function = "crate::util::validate::validate_url"),
+        length(max = 255)
+    )]
+    pub icon_url: Option<String>,
+
+    pub max_scopes: Scopes,
+
+    #[validate(length(min = 1))]
```

Comment:
does this need to be validated to 1? I dont see a big issue with a WIP oauth scope that simply doesnt allow usage yet

Reply:
Are you referring to the `redirect_uris` here? I suppose we could drop the validation, but without any valid redirect URIs you won't be able to do any OAuth stuff. One of the first things it does when starting authorization is heck for a valid redirect uri.

Do you think it would be better to let the user create an 'invalid' application and discover that later when they try to authorize? That seems maybe worse to me but I'm not sure, what do you think?

Reply:
Yeah I am. 
I dont think this is a big deal either way just thinking out loud here. I definitely found use in incomplete WIP oauth stuff when I was doing Ory, but I could also see the value in making it harder for the user to make mistakes. Happy either way

Reply:
Yeah I'm torn too. @Geometrically you have an opinion here?

Reply:
Agree with wyatt it should allow to be empty and just error in the oauth flow

Reply:
Alright cool, removed the validation, should be able to create a client without any redirect URIs now, and you'll still get a nice error message later telling you that that's why it's not working